### ISSUE TITLE
vi (vim) mode / Vim keybindings

### ISSUE THREAD

**Issue by vgel** [warpdotdev/Warp#159]
### Describe the solution you'd like?

Use neovim to drive vim keybindings in the command editor.

### Is your feature request related to a problem? Please describe.

I use Vim keybindings in ZSH (via https://github.com/jeffreytse/zsh-vi-mode) on my home machine. I haven't gotten around to setting it up on my work laptop yet, because it's painful, slow, and buggy. It tries to emulate vim using shell scripts, which doesn't work very well. It'd be really neat to leverage the native editor experience in Warp to provide a near-flawless vim mode like [vscode-neovim](https://github.com/asvetliakov/vscode-neovim) does for vscode.

### Additional context

* https://github.com/jeffreytse/zsh-vi-mode
* https://github.com/asvetliakov/vscode-neovim
* https://github.com/KillTheMule/nvim-rs

**Comment by manuraj17** [comment]
Just curious -- wouldn't it be better if we don't need the dependency on neovim to drive this? 
Ofcourse offloading is better, but then having it native feels better for me. 
Another use case is that I tend to use terminals on machines where neovim is not available -- one example while pairing. So, having an independent mode helps. Just my 2 cents.

**Comment by zheng** [comment]
We definitely want to support some kind of vim mode for the editor. Not sure what our approach will be, but this is a cool project!

When it comes to pairing, we want to support sharing sessions via the web. And one early thought I have is that there's no reason why everyone can't use their own preferred bindings when typing collaboratively (just like in Coderpad)!

**Comment by insufferabblejake** [comment]
Want to throw my hat into the support vim natively without needing neovim or the like.

**Comment by mkomitee** [comment]
Something as basic as what `set -o vi` does in zsh

**Comment by tkennedy1-godaddy** [comment]
This was about 3 minutes of my life trying to figure out why vim keys didn't work in the terminal even though I've got `set -o vi` in my `.bash_profile`.

I have `ctrl-a` bound to other things so emacs keys are pretty hard to use for me.  Oddly if I attach to a tmux session `set -o vi` works again though.

**Comment by elviskahoro** [comment]
@tkennedy1-godaddy Hey Todd,

So the tldr; is the input editor had to built natively to support similar text editing experiences like VS Code. However this also means we'll have to build Vim keybindings ourselves. Once we launch custom keybindings in general, we'll then look to add Vi Mode.

**Comment by hms** [comment]
I know the Warp team has their hands full, but the lack of this feature really makes it very hard to adopt Warp full-time or recommend it as "ready to try".  I hope this feature request will get some love and attention soon as it cuts off a lot of users and I really like Warp and would like to see it be wildly successful.

**Comment by hollyoops** [comment]
Same here very hope Vim can be supported

**Comment by TanGuangZhi** [comment]
now i use zsh in warp to use vim...

**Comment by elviskahoro** [comment]
@gkssjovi @gschnaubelt  Still no solution.
@hms @currenthandle I just checked and it's currently set as low priority. And so I honestly have no sense of what the ETA is. 

> the lack of communications about a roadmap and approximate timing around what's important to me as a user means I've stopped beta testing.

We're trying to figure out this quarter how we can be more transparent about the roadmap.

On the communications front more generally, we're trying to hire more people so that we can improve our responsiveness. After the beta launch we (primarily me) got very overwhelmed with feedback and so sorry that I'm getting to this so late.

**Comment by iryan2** [comment]
Just chiming in as another user that would love to try out this terminal, but am blocked by lack of `set -o vi` (or similar) support.

**Comment by 0x0elliot** [comment]
really excited to see this workout.

**Comment by robertarles** [comment]
> now i use zsh in warp to use vim...

[TanGuangZhi](https://github.com/TanGuangZhi) What do you mean by this exactly. Are you able to get vim keybindings? I run oh-my-zsh, and `bindkeys -v` works for me in iterm, but I'm not getting the vim keybindings when using Warp.

**Comment by jabalsad** [comment]
Just adding my +1. I love everything that warp appears to offer, but vim keybindings are non-negotiable for me.

**Comment by clarkezone** [comment]
Adding a +1.. as much as warp seems nice can't use it without vim keybindings.

**Comment by gmolinart** [comment]
Same here Adding a +1.. can't use it without vim keybindings.

**Comment by programmatio** [comment]
+1

**Comment by wazery** [comment]
+1

**Comment by exsesx** [comment]
+1

**Comment by RobinMaas95** [comment]
+1

**Comment by copalco** [comment]
+1

**Comment by patrick-elmquist** [comment]
Don't care for Neovim but basic bindings like `set -o vi` would be awesome, this together with the login requirement are the only things bugging me with Warp

**Comment by ahayman** [comment]
> now i use zsh in warp to use vim...

How do you do this?  Warp is using zsh as far as I can tell, but vim binding still don't work (even though they work in iTerm & Terminal)

**Comment by exsesx** [comment]
I use fish, by the way, and so far, I haven't had any problems with the vi mode in iTerm2. I used it a lot. Now I've switched to Warp, and I like it even more, but It would've been nice to have that little vi gap closed. Warp would be the best option for me.

**Comment by ahayman** [comment]
Oh, okay. I've primarily use zsh with iTerm, and I love the vi binding for it. I'd love to move to Warp, but missing those bindings is tough.

**Comment by run2xs** [comment]
Lack of vim command line mode is a non-starter for me.  Please add.

**Comment by chuck-hilyard** [comment]
nice product, but vim command line mode is personally paramount.

**Comment by Zamua** [comment]
+1 more for vi mode

**Comment by laz-** [comment]
Just installed warp, stepped through the nux, hit the settings, and I'm here right after.

Taking away the shell's vi input mode without a suitable replacement is a non-starter.

**Comment by juancampa** [comment]
This is a deal-breaker for me too unfortunately. Can't fight muscle memory

**Comment by sifu** [comment]
+1 more for vi mode

**Comment by misterT84** [comment]
+1 more for vi mode

**Comment by jabalsad** [comment]
As a reminder - please 👍  the first post in this issue if you would like your +1 to be counted.

**Comment by ishaikh123** [comment]
+1 more for vim key bindings

**Comment by AndreasPB** [comment]
+1 for vi mode - It would make the switch to Warp a very easy decision

**Comment by mquinnv** [comment]
i've managed to make my brain sorta work now that we can reassign shortcuts by thinking of ctrl/alt as enabling normal mode and binding movement with ctrl and delete with alt. so ctrl+w/b moves and alt+w/b deletes. it is better than the alternative of using arrow keys or learning the gross shortcuts that i won't use anywhere else. not a solution, but stopping me from dumping Warp for the moment. i can't help but feel like i'm dirtying myself with emacs philosophy though

**Comment by zuxfoucault** [comment]
+1 for vim-mode

**Comment by ryboe** [comment]
+1

**Comment by hushyael** [comment]
+1

**Comment by chrismerck** [comment]
Only way I could use this is with vi keybindings. Not having them is going back to the stone age.

**Comment by danrasmuson** [comment]
+1

**Comment by philFernandez** [comment]
I installed Warp back when it first came out, played with it a bit, and then closed it because there were no vim keybindings. I just now decided to open it back up, figuring that vim keybindings would _have_ to be there by now, but nope. This time I just went ahead and trashed the app off my computer. I'll probably never think about it again. It's a shame because I really like the concept, and all of the cool things they've done here, but I just can't work in a terminal without vim keybindings. Its a non sequitur.

**Comment by dieggoluis** [comment]
+1

**Comment by kevinzeidler** [comment]
+1. Warp terminal is easily one of the coolest things I discovered in 2022, and it's been a joy to watch it evolve. It was already the sleekest, most ergonomic terminal emulator I'd ever used when I found it back in May, and with each update it's gotten better and better. But, being a Vimmer at heart, it's been weird to have no Vim keybindings in the command editor. I'd love to see this feature added, it would add so much and it would make it easier for folks to switch to Warp terminal and keep their muscle memory intact.

**Comment by currenthandle** [comment]
I'm just checking in here. Since this issue was opened, over two years ago, I check it all the time. Presently, there are 134 likes on the original post. There are dozens of bumps/+1s, in fact, every time I look at it seems someone else bumped it only a few days prior (or less).

Most frustrating is that the last response from the Warp team was @elviskahoro's response (back in July!!) to my first comment and the content of his statement was that the Warp team was going to be more transparent about the road map in the next quarter. Well, it's been 6 months now, and nothing.

I can only speak for myself, but I imagine that, like me, everyone keeping this thread warm appreciates the work the Warp team is doing to improve the terminal experience. I see Warp adoption in more and more places: https://star-history.com/#warpdotdev/Warp&Date and I only grow more perplexed over time by the fact that this feature still hasn't been integrated.

I know that Vim is amongst the most widely (if not the most) used and loved CLI-based text editors in the world. When I became fluent in Vim and subsequently learned that I could bring these key bindings to my terminal, my productivity and quality of experience grew exponentially overnight. While Warp brings a lot to the table in terms of improvements over the other terminals on the scene, this doesn't make up for what we get by having Vim keybindings in the terminal and we are many.

I am not a desktop app developer nor a Rust dev, although I'm starting to play around a bit :) but it's hard for me to imagine that this feature could be that difficult to implement. Again, I'm a noob here but some communication on this matter from the team would be greatly appreciated! Is there a blocking issue we don't know about? Maybe the community can offer some support here. Or have Vim keybindings been de-prioritized for a long backlog of P0 features and bugs? Again this is hard for me to believe because Warp does some much and runs so smoothly at the time of writing. Maybe more active discussion channels between the community and the Warp team would leave us all more informed on what the community values.

@elviskahoro How are "custom keybindings in general" coming along? https://github.com/warpdotdev/Warp/issues/159#issuecomment-1002917367

Has the Warp team intentionally decided not to ever implement this feature for one reason or another? If so please let us know. 

I'd be happy to organize or take on working on a plugin or forking the Warp repo. If anyone else is eager to work on this let me know. I propose Varp as an obvious name choice.

In conclusion, it's plain to see that Warp is great! But the (massive) Vim gang is hamstrung by the lack of keybindings. Don't you want us Warp team :( We sure want you. Please let us know what's going on or if there's any way we can help. To the community, if anyone wants to take this on, I'm sure we'd all be eternally grateful. I'm happy to as well and if anyone wants to join up or give me some tips on where to look or how to get started that'd be awesome. 🙏🧘‍♂️

**Comment by currenthandle** [comment]
I'm just checking in here. Since this issue was opened, over two years ago, I check it all the time. Presently, there are 134 likes on the original post. There are dozens of bumps/+1s, in fact, every time I look at it seems someone else bumped it only a few days prior (or less).

Most frustrating is that the last response from the Warp team was @elviskahoro's response (back in July!!) to my first comment and the content of his statement was that the Warp team was going to be more transparent about the road map in the next quarter. Well, it's been 6 months now, and nothing.

I can only speak for myself, but I imagine that, like me, everyone keeping this thread warm appreciates the work the Warp team is doing to improve the terminal experience. I see Warp adoption in more and more places: https://star-history.com/#warpdotdev/Warp&Date and I only grow more perplexed over time by the fact that this feature still hasn't been integrated.

I know that Vim is amongst the most widely (if not the most) used and loved CLI-based text editors in the world. When I became fluent in Vim and subsequently learned that I could bring these key bindings to my terminal, my productivity and quality of experience grew exponentially overnight. While Warp brings a lot to the table in terms of improvements over the other terminals on the scene, this doesn't make up for what we get by having Vim keybindings in the terminal and we are many.

I am not a desktop app developer nor a Rust dev, although I'm starting to play around a bit :) but it's hard for me to imagine that this feature could be that difficult to implement. Again, I'm a noob here but some communication on this matter from the team would be greatly appreciated! Is there a blocking issue we don't know about? Maybe the community can offer some support here. Or have Vim keybindings been de-prioritized for a long backlog of P0 features and bugs? Again this is hard for me to believe because Warp does some much and runs so smoothly at the time of writing. Maybe more active discussion channels between the community and the Warp team would leave us all more informed on what the community values.

@elviskahoro How are "custom keybindings in general" coming along? https://github.com/warpdotdev/Warp/issues/159#issuecomment-1002917367

Has the Warp team intentionally decided not to ever implement this feature for one reason or another? If so please let us know.

I'd be happy to organize or take on working on a plugin or forking the Warp repo. If anyone else is eager to work on this let me know. I propose Varp as an obvious name choice.

In conclusion, it's plain to see that Warp is great! But the (massive) Vim gang is hamstrung by the lack of keybindings. Don't you want us Warp team :( We sure want you. Please let us know what's going on or if there's any way we can help. To the community, if anyone wants to take this on, I'm sure we'd all be eternally grateful. I'm happy to as well and if anyone wants to join up or give me some tips on where to look or how to get started that'd be awesome. 🙏🧘

**Comment by currenthandle** [comment]
I'm just checking in here. Since this issue was opened, over two years ago, I check it all the time. Presently, there are 134 likes on the original post. There are dozens of bumps/+1s, in fact, every time I look at it seems someone else bumped it only a few days prior (or less).

Most frustrating is that the last response from the Warp team was @elviskahoro's response https://github.com/warpdotdev/Warp/issues/159#issuecomment-1174601921 (back in July!!) to my first comment and the content of his statement was that the Warp team was going to be more transparent about the road map in the next quarter. Well, it's been 6 months now, and nothing.

I can only speak for myself, but I imagine that, like me, everyone keeping this thread warm appreciates the work the Warp team is doing to improve the terminal experience. I see Warp adoption in more and more places: https://star-history.com/#warpdotdev/Warp&Date and I only grow more perplexed over time by the fact that this feature still hasn't been integrated.

I know that Vim is amongst the most widely (if not the most) used and loved CLI-based text editors in the world. When I became fluent in Vim and subsequently learned that I could bring these key bindings to my terminal, my productivity and quality of experience grew exponentially overnight. While Warp brings a lot to the table in terms of improvements over the other terminals on the scene, this doesn't make up for what we get by having Vim keybindings in the terminal and we are many.

I am not a desktop app developer nor a Rust dev, although I'm starting to play around a bit :) but it's hard for me to imagine that this feature could be that difficult to implement. Again, I'm a noob here but some communication on this matter from the team would be greatly appreciated! Is there a blocking issue we don't know about? Maybe the community can offer some support here. Or have Vim keybindings been de-prioritized for a long backlog of P0 features and bugs? Again this is hard for me to believe because Warp does some much and runs so smoothly at the time of writing. Maybe more active discussion channels between the community and the Warp team would leave us all more informed on what the community values.

@elviskahoro How are "custom keybindings in general" coming along? https://github.com/warpdotdev/Warp/issues/159#issuecomment-1002917367

Has the Warp team intentionally decided not to ever implement this feature for one reason or another? If so please let us know.

I'd be happy to organize or take on working on a plugin or forking the Warp repo. If anyone else is eager to work on this let me know. I propose Varp as an obvious name choice.

In conclusion, it's plain to see that Warp is great! But the (massive) Vim gang is hamstrung by the lack of keybindings. Don't you want us Warp team :( We sure want you. Please let us know what's going on or if there's any way we can help. To the community, if anyone wants to take this on, I'm sure we'd all be eternally grateful. I'm happy to as well and if anyone wants to join up or give me some tips on where to look or how to get started that'd be awesome. 🙏🧘

**Comment by currenthandle** [comment]
I'm just checking in here. Since this issue was opened, over a years ago, I check it all the time. Presently, there are 134 likes on the original post. There are dozens of bumps/+1s, in fact, every time I look at it seems someone else bumped it only a few days prior (or less).

Most frustrating is that the last response from the Warp team was @elviskahoro's response https://github.com/warpdotdev/Warp/issues/159#issuecomment-1174601921 (back in July!!) to my first comment and the content of his statement was that the Warp team was going to be more transparent about the road map in the next quarter. Well, it's been 6 months now, and nothing.

I can only speak for myself, but I imagine that, like me, everyone keeping this thread warm appreciates the work the Warp team is doing to improve the terminal experience. I see Warp adoption in more and more places: https://star-history.com/#warpdotdev/Warp&Date and I only grow more perplexed over time by the fact that this feature still hasn't been integrated.

I know that Vim is amongst the most widely (if not the most) used and loved CLI-based text editors in the world. When I became fluent in Vim and subsequently learned that I could bring these key bindings to my terminal, my productivity and quality of experience grew exponentially overnight. While Warp brings a lot to the table in terms of improvements over the other terminals on the scene, this doesn't make up for what we get by having Vim keybindings in the terminal and we are many.

I am not a desktop app developer nor a Rust dev, although I'm starting to play around a bit :) but it's hard for me to imagine that this feature could be that difficult to implement. Again, I'm a noob here but some communication on this matter from the team would be greatly appreciated! Is there a blocking issue we don't know about? Maybe the community can offer some support here. Or have Vim keybindings been de-prioritized for a long backlog of P0 features and bugs? Again this is hard for me to believe because Warp does some much and runs so smoothly at the time of writing. Maybe more active discussion channels between the community and the Warp team would leave us all more informed on what the community values.

@elviskahoro How are "custom keybindings in general" coming along? https://github.com/warpdotdev/Warp/issues/159#issuecomment-1002917367

Has the Warp team intentionally decided not to ever implement this feature for one reason or another? If so please let us know.

I'd be happy to organize or take on working on a plugin or forking the Warp repo. If anyone else is eager to work on this let me know. I propose Varp as an obvious name choice.

In conclusion, it's plain to see that Warp is great! But the (massive) Vim gang is hamstrung by the lack of keybindings. Don't you want us Warp team :( We sure want you. Please let us know what's going on or if there's any way we can help. To the community, if anyone wants to take this on, I'm sure we'd all be eternally grateful. I'm happy to as well and if anyone wants to join up or give me some tips on where to look or how to get started that'd be awesome. 🙏🧘

**Comment by dannyneira** [comment]
Thanks for your detailed feedback @currenthandle, we have several big issues/requests on the [roadmap](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3ARoadmap) and will add the label/tag to this issue when it's being worked on. Given we've implemented our own modern input editor, vim keybinds would have to be done from scratch to work in Warp. We appreciate your patience as we try and knock out these major requests.

**Comment by dannyneira** [comment]
Thanks for the feedback @currenthandle, I agree this is a big request, but given we've implemented our own Input Editor into the terminal, we'll need some time to work on this feature. We do have quite a few big ticket items on the [roadmap](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3ARoadmap), and will add the label to this request once it's being worked on.

**Comment by dannyneira** [comment]
Thanks for the feedback @currenthandle, I agree this is a big request, but given we've implemented our own Input Editor into the terminal, we'll need some time to work on this feature. We do have quite a few big-ticket items on the [roadmap](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3ARoadmap), and will add the label to this request once it's being worked on.

**Comment by hms** [comment]
Danny,

You seem to be missing the point.  Again. Everyone understands how big a job building warp is.  Everyone understands you have a roadmap that must meet your business commitments and that your roadmap might not match our desires.

What the warp team seams to not understand is how the lack of communication on a topic that is so demonstrably important (size and age of thread.....) burns good will with your beta testers. No one is asking you or the team to change your priorities to meet ours.  All we asked, and waited patiently for, was a realistic response or roadmap timeline.   

Given how the Warp team has so clearly deprioritized a feature that is critical to my workflow, I dropped out of the beta.  That's ok for both of us -- I'm not a good tester without VIM keyboard bindings and you clearly aren't catering to people with similar needs as demonstrated by the lack of that functionality.

That being said, I will never become a Warp user in the future unless I can be sure it has 100% (not 99% -- not good enough) of everything I think I need since this year has demonstrated that a lack of communication is a feature, not a bug.  This thread and the length of time I've waited for any kind of useful feedback has turned me off as a user.

I wish you and the team good luck. 

hms

**Comment by pr3d4t0r** [comment]
+1 for Vim mode
-1 for NeoVim (or Vim) dependency - it should be built-in

bash and zsh already have Vim bindings (`set -o vi`) so maybe the solution would be to try to bind with the shell itself rather than with the editor?  bash's Vim bindings support gets a lot of flack because it "supports" visual mode by opening the editor itself -- that doesn't sound like a good idea.

I'll have a look at how Warp bindings are implemented now and then opine or start submitting pull requests.

**Comment by dimon1553** [comment]
+ set -o vi

**Comment by jmorecroft-seek** [comment]
yep, was quite excited about Warp until I realised this didn't work. and out. back to the terminal, nice knowing you Warp.

**Comment by UDtorrey** [comment]
+1 for `set -o vi`; I was excited seeing some colleagues use warp; however my muscle memory wont adjust - i've tried. some tricks you cant teach old dogs.

**Comment by AnotherYeti** [comment]
+1 for this, +2 for using a more modern Kakoune/Helix style option.

**Comment by jsonMartin** [comment]
I'm enjoying using warp - especially the AI command integration, but without `vim` keybindings is a dealbreaker. Using vim keyboard shortcuts to manipulate text directly in the CLI is very helpful.

**Comment by eccomaggio** [comment]
+1 for vim mode

**Comment by icebrg** [comment]
I also can't do without vim mode (however it's supported).

**Comment by guttenbergovitz** [comment]
+1 for mouse bashing...

**Comment by guttenbergovitz** [comment]
anyway... using tmux with zsh on top of warp, but there is something fundamentally wrong with my approach I reckon...

**Comment by DieracDelta** [comment]
+1 for vim mode

**Comment by lpxxn** [comment]
+1 for vim mode

**Comment by horserad** [comment]
+1 for vim mode

**Comment by haveanicedavid** [comment]
+1 for the "Wanna use warp but no vi movements is a non-starter" gang 😄

**Comment by bdtomlin** [comment]
I hate to be a +1 guy, but here I am. I just found out about warp today and installed it. I kind of love it. I'm going to continue to try using it without vim key bindings, but I'm not sure how long I can last. Lack of basic vim keybindings is very painful.

**Comment by hmasood-px** [comment]
+1 for vim mode!!!

**Comment by GregoryConrad** [comment]
FWIW, it would be nice if you did not embed neovim, but rather had the system generic enough that you could add keybindings for different types of editors (like Helix or Emacs) too. That would be fantastic for those of us that don't use vim.

**Comment by benlieb** [comment]
@GregoryConrad vi mode is just a mode, and wouldn't preclude any other mode. 

`set -o vi`

**Comment by Manubi** [comment]
maybe add it to the documentation that you don't support it yet.
Probably changing back to iterm2.

**Comment by Lincest** [comment]
+1 for vim mode

**Comment by brennschlus** [comment]
it's already April 2023, any news or updates on this?

**Comment by Manubi** [comment]
just to add it, it would be nice to also have the vim commands to move between panels and windows. such as tmux.

**Comment by Tomotoes** [comment]
vi mode, please

**Comment by bryant-finney** [comment]
+1 for vi mode (it's the only reason I'm still using iTerm2 as my daily driver)

**Comment by BeauSLM** [comment]
+1 for vi (or helix-type) mode

**Comment by stounfo** [comment]
+1 for vi mode

**Comment by Vikaiv** [comment]
+1 for vi mode

**Comment by samolin** [comment]
+1 for vim mode

**Comment by RPDavlyatshin** [comment]
+1 for vi mode

**Comment by kevinjruffe** [comment]
+1 for vi mode

**Comment by djmango** [comment]
+1

**Comment by marc-brouard** [comment]
+1

**Comment by rm374** [comment]
+1 as well - downloaded warp recently and while I love some of the features it offers right out of the box, my muscle memory is too deeply tied to vi mode. Would be awesome if it was supported here as well.

**Comment by sashaaldrick** [comment]
+11 for vi mode, I just switched to warp, but not sure I can keep using it without vim binds, or at least `bindkey` support. Thanks.

**Comment by dannyneira** [comment]
Hey Folks, This feature is on the roadmap and actively being worked on (no official eta). We'll post any updates on this thread once we have more to share! 🤞

**Comment by Meronkha** [comment]
+1 on this feature

**Comment by Sikian** [comment]
I really like the idea of warp, but unfortunately can't use it without key bindings. I'll keep an eye on this issue to see when I can come back.

**Comment by currenthandle** [comment]
> Hey Folks, This feature is on the roadmap and actively being worked on (no official eta). We'll post any updates on this thread once we have more to share! 🤞 Sneak peek below:
> 
> ![CleanShot 2023-07-18 at 11 14 31](https://user-images.githubusercontent.com/16809145/254343755-3ae3790e-70fe-4f5a-904e-93d4504d1c1f.png)

Hey @dannyneira happy August! Any way we can get an update?

**Comment by michaelmcmillan** [comment]
+1 for vi mode!

Also, this would suffice for me, as it allows me drop into vi with the current command:

```
set -o vi
```

**Comment by Amarlanda** [comment]
excited to see this being worked on +1

**Comment by evelant** [comment]
Excited to see this in the works. Can't wait to try it!

**Comment by juancampa** [comment]
This would be a game changer if it supported text objects (e.g. `iw`, `aw`) which none of the shells I've used do

**Comment by dannyneira** [comment]
Hey folks, we're making good progress on vi mode, adding more supported shortcuts every week! We want to make sure it's just right before we launch, so we appreciate your patience while we work on this highly requested feature. (@juancampa It will support text object shortcuts 🤌 )

**Comment by benlieb** [comment]
Is it not possible/easyish to simply have an embedded instance of nvim? I can even do it in Chrome.

**Comment by jeffschwartz** [comment]
Both Bash and ZSH supports vi mode, Bash directly via .profile, ZSH via ZSH plugins. These work in iTerm2 very well. Couldn't Warp tap into BASH and ZSH to do similarly?

**Comment by jeffschwartz** [comment]
Why create what the underlying shells mostly support already,  e.g. Bash, ZSH. These allow vi integration either directly or through plugins. I don't believe this should be a terminal implementation. Rather, Warp just needs to integrate with the shell to support this feature.

**Comment by lucieleblanc** [comment]
Hello fellow Vimmers, Lucie (@lucieleblanc) and Andy (@acarl005) here from Warp's engineering team! We've been working on Vim keybindings, and as of the latest release you can now use them in Warp's input editor 🎉

To enable them, navigate to `Settings` → `Features` → `Editor`, or search in the command palette with `cmd-p`, and turn on "Edit commands with Vim keybindings":
<img width="492" alt="vim-keybindings-beta-settings-toggle" src="https://github.com/warpdotdev/Warp/assets/14223323/0ea22dab-772e-4a8a-85be-d585bd5fc897">

See the documentation [here](docs.warp.dev/features/editor/vim) for a list of supported keybindings. 

Vim keybindings are currently in beta, and we want to hear your feedback. We've been using this internally, but there are certainly bugs and use cases we haven't encountered yet. Although this main issue thread will be closed, please open new issues in this repo and upvote (:thumbsup:) existing Vim-related issues to help us prioritize them.

For those interested, we implemented Vim keybindings natively as a finite-state automaton (FSA) that plugs into Warp's input editor. We're hoping to [write a blog post about it](https://github.com/warpdotdev/Warp/issues/3735), and maybe [open-source our Vim FSA implementation](https://github.com/warpdotdev/Warp/issues/3736) in its own crate – please upvote those issues if they would be of interest to you.

**Comment by djmango** [comment]
This is great! Will try switching over to warp now from iterm2 full time. Is this bug reported? Nvim/lvim has borders around it
<img width="1512" alt="image" src="https://github.com/warpdotdev/Warp/assets/28496988/0c760c4f-56f2-4b5c-bc92-ca171c1cb554">

**Comment by iryan2** [comment]
Legends!! Thank you!

**Comment by benlieb** [comment]
Great news that there is some movement in the pond. 

A question. Since this is using a gui toggle, If I have `set -o vi` in an rc file it will have no effect?

**Comment by acarl005** [comment]
@benlieb Correct, setting the vi mode in the shell (such as with `set -o vi`) currently has no effect in Warp. We are still working out the UX for interfacing with the shell's setting.

**Comment by nikisix** [comment]
Highlight of my week right here. Can't wait to get back into warp!

**Comment by danrasmuson** [comment]
They have listened and we have been heard!

**Comment by bdtomlin** [comment]
I'm very happy with this implementation. Great job, thanks for listening!

**Comment by apolzon** [comment]
Sorry to pile-on -- the implementation is indeed terrific (I've even been surprised by how thorough the command support is) -- I know that yall are working on better support for non-shell subshells (ie rails console, python, etc.) -- when/if that work gets completed, will the VIM support work there as well?

**Comment by djmango** [comment]
One bug/UX issue I've noticed, when using J/K to scroll command history, it is essentially impossible to edit multiline commands as J/K just goes to the next/prev commands instead of navigating lines of the command. You are only able to edit without running by clicking with a mouse/trackpad, which defeats the purpose of VIM

**Comment by cshintov** [comment]
> One bug/UX issue I've noticed, when using J/K to scroll command history, it is essentially impossible to edit multiline commands as J/K just goes to the next/prev commands instead of navigating lines of the command. You are only able to edit without running by clicking with a mouse/trackpad, which defeats the purpose of VIM

Ya, faced the same. Also `ESC-v` doesnt' work!

There's a workaround though - use `fc` directly.

**Comment by lucieleblanc** [comment]
Thank you all for the feedback! It is much appreciated.

@apolzon Yes, Warp's editor code remains the same in subshells, so I expect that Vim keybindings could work in non-shell REPLs too. However, I'm not sure how they would interact with REPLs that have non-standard bindings built in. Non-shell REPLs are not on Warp's roadmap for the next few months, so I can't promise anything there.

@djmango Currently j/k behave like up/down arrow keys, so the same problem exists outside of vim mode. We're working on finding a better UX for this, but please do open a separate issue for it so we can share and discuss ideas for a resolution!

@cshintov I'm not familiar with `ESC-v`. Could you open a separate issue to describe how you use it in other terminals and/or how you would like to use it in Warp? That would be very helpful!

**Comment by cshintov** [comment]
@djmango anyone interested in this please like and subscribe 😉

command edit and execution feature like vi mode.

https://github.com/warpdotdev/Warp/issues/3795

**Comment by ratbekn** [comment]
+1 on this feature!

**Comment by ayyjohn** [comment]
would love to see the ability for warp to read from my vimrc so that i can use my custom keybindings :(
https://docs.warp.dev/features/editor/vim explicitly states only default mappings are allowed.
